File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        if (root == null)
            return list;
        return inorder(root, list);
    }
    
    private List<Integer> inorder(TreeNode root, List<Integer> list) {
        if (root == null)
            return list;
        inorder(root.left, list);
        list.add(root.val);
        return inorder(root.right, list);
    }
}

File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        TreeNode* cur = root;
        TreeNode* pre = NULL;
        while (cur != NULL) {
            if (cur->left == NULL) {
                res.push_back(cur->val);
                cur = cur->right;
            } else {
                pre = cur->left;
                while (pre->right != NULL)
                    pre = pre->right;
                pre->right = cur;
                TreeNode* temp = cur;
                cur = cur->left;
                temp->left = NULL;
            }
        }
        return res;
    }
};